src/plan.rs

```diff
@@ -0,0 +1,799 @@
+    impl_log_method!(check_older, s: Sequence, -> bool);
+    impl_log_method!(check_after, t: LockTime, -> bool);
+}
+
+impl<T, Pk> AssetProvider<Pk> for T
+where
+    T: Satisfier<Pk>,
+    Pk: MiniscriptKey + ToPublicKey,
+{
+    fn provider_lookup_ecdsa_sig(&self, pk: &Pk) -> bool {
+        Satisfier::lookup_ecdsa_sig(self, pk).is_some()
+    }
+
+    fn provider_lookup_tap_key_spend_sig(&self, _: &Pk) -> Option<usize> {
+        Satisfier::lookup_tap_key_spend_sig(self).map(|s| s.to_vec().len())
+    }
+
+    fn provider_lookup_tap_leaf_script_sig(
+        &self,
+        pk: &Pk,
+        leaf_hash: &TapLeafHash,
+    ) -> Option<usize> {
+        Satisfier::lookup_tap_leaf_script_sig(self, pk, leaf_hash).map(|s| s.to_vec().len())
+    }
+
+    fn provider_lookup_tap_control_block_map(
+        &self,
+    ) -> Option<&BTreeMap<ControlBlock, (bitcoin::ScriptBuf, LeafVersion)>> {
+        Satisfier::lookup_tap_control_block_map(self)
+    }
+
+    fn provider_lookup_raw_pkh_pk(&self, hash: &hash160::Hash) -> Option<bitcoin::PublicKey> {
+        Satisfier::lookup_raw_pkh_pk(self, hash)
+    }
+
+    fn provider_lookup_raw_pkh_x_only_pk(&self, hash: &hash160::Hash) -> Option<XOnlyPublicKey> {
+        Satisfier::lookup_raw_pkh_x_only_pk(self, hash)
+    }
+
+    fn provider_lookup_raw_pkh_ecdsa_sig(
+        &self,
+        hash: &hash160::Hash,
+    ) -> Option<bitcoin::PublicKey> {
+        Satisfier::lookup_raw_pkh_ecdsa_sig(self, hash).map(|(pk, _)| pk)
+    }
+
+    fn provider_lookup_raw_pkh_tap_leaf_script_sig(
+        &self,
+        hash: &(hash160::Hash, TapLeafHash),
+    ) -> Option<(XOnlyPublicKey, usize)> {
+        Satisfier::lookup_raw_pkh_tap_leaf_script_sig(self, hash)
+            .map(|(pk, sig)| (pk, sig.to_vec().len()))
+    }
+
+    fn provider_lookup_sha256(&self, hash: &Pk::Sha256) -> bool {
+        Satisfier::lookup_sha256(self, hash).is_some()
+    }
+
+    fn provider_lookup_hash256(&self, hash: &Pk::Hash256) -> bool {
+        Satisfier::lookup_hash256(self, hash).is_some()
+    }
+
+    fn provider_lookup_ripemd160(&self, hash: &Pk::Ripemd160) -> bool {
+        Satisfier::lookup_ripemd160(self, hash).is_some()
+    }
+
+    fn provider_lookup_hash160(&self, hash: &Pk::Hash160) -> bool {
+        Satisfier::lookup_hash160(self, hash).is_some()
+    }
+
+    fn check_older(&self, s: Sequence) -> bool {
+        Satisfier::check_older(self, s)
+    }
+
+    fn check_after(&self, l: LockTime) -> bool {
+        Satisfier::check_after(self, l)
+    }
+}
+
+/// Representation of a particular spending path on a descriptor. Contains the witness template
+/// and the timelocks needed for satisfying the plan.
+/// Calling `plan` on a Descriptor will return this structure,
+/// containing the cheapest spending path possible (considering the `Assets` given)
+#[derive(Debug, Clone)]
+pub struct Plan {
+    /// This plan's witness template
+    pub(crate) template: Vec<Placeholder<DefiniteDescriptorKey>>,
```

Comment:
In d29c2984e60e41e3872ac23b8b59e8b25fea88a2:

Maybe this was discussed in the other thread, but should we make this generic over `Pk` rather than using a `DefiniteDescriptorKey`?